### PR TITLE
extended tests for jenkins openshift V3 plugin

### PR DESCRIPTION
**Pull request by gabemontero**
@bparees @mfojtik PTAL

### CODE REVIEW

File: test/extended/jenkins/jenkins_plugin.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+	pre := exutil.FixturePath("fixtures", filename)
+	post := exutil.ArtifactPath(filename)
+	err := exutil.VarSubOnFile(pre, post, "PROJECT_NAME", namespace)
+	o.Expect(err).NotTo(o.HaveOccurred())
+	data, err := ioutil.ReadFile(post)
+	o.Expect(err).NotTo(o.HaveOccurred())
+	return data
+}
+
+var _ = g.Describe("jenkins: plugin: run job leverging openshift pipeline plugin", func() {
```
Comment (bparees) [inline test/extended/jenkins/jenkins_plugin.go]: leveraging

File: test/extended/jenkins/jenkins_plugin.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	var hostPort string
+
+	g.BeforeEach(func() {
+
+		g.By("set up policy for jenkins jobs")
+		err := oc.Run("policy").Args("add-role-to-user", "edit", "system:serviceaccount:"+oc.Namespace()+":default").Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("kick off the build for the jenkins ephermeral and application templates")
+		jenkinsEphemeralPath := exutil.FixturePath("fixtures", "jenkins-ephemeral-template.json")
```
Comment (bparees) [inline test/extended/jenkins/jenkins_plugin.go]: if these two templates aren't changed from what's in examples, just reference them in examples instead of making a copy.  that way the test also ensures we don't break the example template files.

File: test/extended/jenkins/jenkins_plugin.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+		err = exutil.NewApp(jenkinsEphemeralPath, oc)
+		o.Expect(err).NotTo(o.HaveOccurred())
+		jenkinsApplicationPath := exutil.FixturePath("fixtures", "application-template.json")
+		err = exutil.NewApp(jenkinsApplicationPath, oc)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("tag the app image stream")
+		err = exutil.TagImageStream("nodejs-010-centos7", "openshift/nodejs-010-centos7:latest", oc)
```
Comment (bparees) [inline test/extended/jenkins/jenkins_plugin.go]: i think all the default imagestreams are already created when running extended tests.....not clear why this is needed?

File: test/extended/jenkins/jenkins_plugin.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+			immediateInteractionWithJenkins(fmt.Sprintf("http://%s/createItem?name=test-plugin-job", hostPort), "POST", bytes.NewBuffer(data), 200)
+
+			g.By("make http request to kick off build")
+			immediateInteractionWithJenkins(fmt.Sprintf("http://%s/job/test-plugin-job/build?delay=0sec", hostPort), "POST", nil, 201)
+
+			// the build and deployment is by far the most time consuming portion of the test jenkins job;
+			// we leverage some of the openshift utilities for waiting for the deployment before we poll
+			// jenkins for the sucessful job completion
+			g.By("waiting for frontend deployment as sign that the build has finished")
+			err := exutil.WaitForADeployment(oc.KubeREST().ReplicationControllers(oc.Namespace()), "frontend", exutil.CheckDeploymentCompletedFn, exutil.CheckDeploymentFailedFn)
```
Comment (bparees) [inline test/extended/jenkins/jenkins_plugin.go]: how long does this wait before giving up?  looks like it might wait forever which would be bad for fast-failing.

File: test/extended/jenkins/jenkins_plugin.go (hunk 3 above)
Comment (gabemontero) [inline test/extended/jenkins/jenkins_plugin.go]: the equivalent command has been in my jenkins env startup scripts for the longest time, which were based off of the jenkins instructions; but in reviewing the instructions now, I don't see it there ... so either the instructions have been cleaned up, or I mistakenly put thing in back then ... don't remember at this point ;-)

In any event, I'll try with this removed, and see what happens

File: test/extended/jenkins/jenkins_plugin.go (hunk 4 above)
Comment (gabemontero) [inline test/extended/jenkins/jenkins_plugin.go]: yeah, perhaps it depends on the overall ginkgo timeout, or the watch has a TO, but otherwise, I agree it seems like it would wait forever ... WaitForAnImageStream seems similar; it is leveraged in some other test cases;

According to `git blame`, @stevekuznetsov last manipulated these methods .. any insight on when those were expected (or not expected) to timeout @stevekuznetsov ?

Otherwise, @bparees I could leverage WaitForABuild in the new test case ... it is not quite as far along the build flow, but still gives us a head start for subsequent polling

File: test/extended/jenkins/jenkins_plugin.go (hunk 3 above)
Comment (gabemontero) [inline test/extended/jenkins/jenkins_plugin.go]: i worked without the `oc tag` ... i'll leave the utility method in the utils package in there unless you disagree - seems like it could be useful down the road for somebody

File: test/extended/jenkins/jenkins_plugin.go (hunk 4 above)
Comment (bparees) [inline test/extended/jenkins/jenkins_plugin.go]: the watch timeout won't matter since it'll just re-list and then start a new watch if the watch times out.

the right thing to do is probably do make WaitForADeployment take a duration, or at least make it implicitly include one like WaitForABuild does.

File: test/extended/jenkins/jenkins_plugin.go (hunk 3 above)
Comment (bparees) [inline test/extended/jenkins/jenkins_plugin.go]: yeah that seems fine.

File: test/extended/jenkins/jenkins_plugin.go (hunk 4 above)
Comment (gabemontero) [inline test/extended/jenkins/jenkins_plugin.go]: want to make sure I reconcile your "re-list" comment correctly - you mean the logic in `WaitForADeployment` with  this code:
`val, ok := <-w.ResultChan()`
`if !ok {`
`   // reget and re-watch`
`   break`
where the watch timing out means w.ResultChan() exits, and the starting of the new watch is in `WaitForADeployment`, not some magic withing w.ResultChan(), correct?

**Comment by gabemontero** [comment]
@bparees updates based on 1st set of comments

File: test/extended/jenkins/jenkins_plugin.go (hunk 4 above)
Comment (bparees) [inline test/extended/jenkins/jenkins_plugin.go]: right, just saying that when w.ResultChan fails, the existing WaitForADeployment logic is going to go back to the top of the loop, do a fresh list of the resources, and then start a new watch on the resources.  so it's not going to return based on the watch timing out.

File: test/extended/util/framework.go [hunk 5]
```diff
@@ -189,22 +189,22 @@ var CheckImageStreamTagNotFoundFn = func(i *imageapi.ImageStream) bool {
 func WaitForADeployment(client kclient.ReplicationControllerInterface,
```
Comment (bparees) [inline test/extended/util/framework.go]: I think i would have done this differently.... either:

1) just check the start time and calculate and end time, and then if the end time has passed at the top of the outer loop
2) get rid of the the Watch entirely and just wrap the list() in a wait.Poll check (which is basically what WaitForABuild does).

with the existing logic it's pretty hard to follow how long it's really going to end up waiting, based on expiration of watches.  also 3 minutes between polls seems like a lot.

File: test/extended/jenkins/jenkins_plugin.go (hunk 4 above)
Comment (stevekuznetsov) [inline test/extended/jenkins/jenkins_plugin.go]: I believe the `WaitFor` methods were never used for something that was expected to fail so the timeouts on them were those imposed by Ginkgo on the rest itself.

File: test/extended/util/framework.go (hunk 5 above)
Comment (gabemontero) [inline test/extended/util/framework.go]: went with option 1) ... will push in a sec

File: test/extended/jenkins/jenkins_plugin.go (hunk 4 above)
Comment (bparees) [inline test/extended/jenkins/jenkins_plugin.go]: That doesn't strike me as sound reasoning... These are tests, everything
should be expected to potentially fail...

Ben Parees | OpenShift

Comment (stevekuznetsov) [inline test/extended/jenkins/jenkins_plugin.go]: Quite a lot of our caching unit tests do exactly the same thing, where the native timeout given by `go test` is used as the threshold. The tests mostly predate me so I assumed we used the same approach here.

**Comment by gabemontero** [comment]
[testonlyextended]

File: test/extended/util/image_helpers.go [hunk 6]
```diff
@@ -97,3 +97,15 @@ func CreateResource(jsonFilePath string, oc *CLI) error {
 	err := oc.Run("create").Args("-f", jsonFilePath).Execute()
 	return err
 }
+
+//NewApp performs an oc new-app invocation from the supplied json file
+func NewApp(jsonFilePath string, oc *CLI) error {
+	err := oc.Run("new-app").Args(jsonFilePath).Execute()
+	return err
```
Comment (mfojtik) [inline test/extended/util/image_helpers.go]: just return the oc.Run here :-)

File: test/extended/util/image_helpers.go [hunk 7]
```diff
@@ -97,3 +97,15 @@ func CreateResource(jsonFilePath string, oc *CLI) error {
...
+}
+
+//TagImageStream performs and oc tag invocation with the provided parameters
+func TagImageStream(source, destination string, oc *CLI) error {
+	err := oc.Run("tag").Args(source, destination, "--config", KubeConfigPath()).Execute()
+	return err
```
Comment (mfojtik) [inline test/extended/util/image_helpers.go]: same here

File: test/extended/util/image_helpers.go (hunk 6 above)
Comment (mfojtik) [inline test/extended/util/image_helpers.go]: is this worth to have special function? I think reading the code will tell you what this does.... having NewApp() makes you go find the definition if something explode :)

File: test/extended/util/image_helpers.go (hunk 7 above)
Comment (gabemontero) [inline test/extended/util/image_helpers.go]: in hindsight yes that seem obvious (i think i was originally expecting something more elaborate in those methods) ;-)

getting rid of both

**Comment by gabemontero** [comment]
[testonlyextended][extended:core]

**Comment by gabemontero** [comment]
[testonlyextended][extended:core]

**Comment by gabemontero** [comment]
[testonlyextended][extended:core(jenkins)]

**Comment by gabemontero** [comment]
hack/install-assets.sh failed ... never got to the extended tests in the extended test run

**Comment by gabemontero** [comment]
[testonlyextended][extended:core(jenkins)]

File: test/extended/util/framework.go [hunk 8]
```diff
@@ -189,7 +189,9 @@ var CheckImageStreamTagNotFoundFn = func(i *imageapi.ImageStream) bool {
...
 	name string,
 	isOK, isFailed func(*kapi.ReplicationController) bool) error {
-	for {
+	startTime := time.Now()
+	endTime := startTime.Add(3 * time.Minute)
```
Comment (bparees) [inline test/extended/util/framework.go]: bump this up to something like 15 or 20 minutes.  we don't want to introduce flakes if we have existing extended tests that take a long time waiting for a deployment for some reason.

i'm also surprised 3 minutes is long enough for you test, given that it basically has to get through the entire imagebuild+deploy step in 3 minutes to pass. 

(3 minutes or the watch timeout, whichever is longer.)

Comment (gabemontero) [inline test/extended/util/framework.go]: Will do - it worked for me locally but i did not measure how long it
typically took

Comment (bparees) [inline test/extended/util/framework.go]: I'm guessing the watch doesn't get closed because there's not much etcd
change log activity, so you don't come back to the top of the loop, so the
3 minute limit doesn't come into play in your particular scenario.

Ben Parees | OpenShift

**Comment by gabemontero** [comment]
the test only extended keep failing in a early, just after build-go.sh completes successfully, in a weird way, way before it gets to the extended test.

tried refreshing the branch; if still fails, going to remove the existing test flags and just try a generic test job.

**Comment by gabemontero** [comment]
ok ... it looks like it got to try and run the extended test, but it seems to fail immediately with an error that is not obvious to me at first blush:

Running TEST_REPORT_DIR=/tmp/openshift-extended-tests/junit/extended test/extended/core.sh --ginkgo.focus="jenkins"...
package k8s.io/kubernetes/pkg/util
	imports github.com/davecgh/go-spew/spew
	imports github.com/docker/libcontainer/cgroups/fs
	imports github.com/docker/docker/pkg/mount
	imports github.com/docker/docker/pkg/units
	imports github.com/docker/docker/pkg/units
	imports github.com/docker/docker/pkg/units: cannot find package "github.com/docker/docker/pkg/units" in any of:
	/usr/lib/golang/src/github.com/docker/docker/pkg/units (from $GOROOT)
	/data/src/github.com/docker/docker/pkg/units (from $GOPATH)

I'm going to see about clearing all the prior test flags, see if I can get full fledged test going.

**Comment by mfojtik** [comment]
@gabemontero why do you import "github.com/docker/docker/pkg/units" ?

**Comment by mfojtik** [comment]
OK this seems like kubernetes imports that package... that is weird, do you have godeps setup properly?

**Comment by gabemontero** [comment]
@mfojtik I don't import "github.com/docker/docker/pkg/units" .... as far as i know i do, I did not make any godeps changes as part of this PR ... in any event, wouldn't the test job on jenkins be the one that pulls the various godeps?

**Comment by mfojtik** [comment]
@gabemontero I just ran that command and it worked for me, that means that you might have something wrong with your environment... can you check the value of GOPATH?

**Comment by gabemontero** [comment]
[test]

**Comment by bparees** [comment]
lgtm, waiting for extended test framework to be fixed up.

**Comment by gabemontero** [comment]
[testonlyextended][extended:core(jenkins)]

**Comment by gabemontero** [comment]
back to testing, having pulled in pr https://github.com/openshift/origin/pull/6558 while it is waiting to merge, and uncovered a timing issue panic in the wait for jenkins to come up logic ... pushing an update for retest shortly.

**Comment by gabemontero** [comment]
ok ... the new extended test just passed !! ... Going to push the back out of the 6558 fix, confirm things are still good

**Comment by gabemontero** [comment]
OK that's two straight successful extended test runs.  And the travis hiccup appears to be some sort of go 1.4 flake (it works in 1.5).

@bparees IMO - this is ready to hit the merge button on - PTAL at your convenience / discretion

**Comment by gabemontero** [comment]
whoaaa .... very cool tool;  thx for the tip re: gvm !! repro'ed, working on fix ... glad my speculation regarding go versions and DeepEquals turned out to be relevant :-)

File: test/extended/jenkins/jenkins_plugin.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+	})
+
+	if err != nil {
+		return fmt.Errorf("got error %v waiting on uri %s with verificationString %s and last set of console logs %s", err, uri, verificationString, consoleLogs)
+	} else {
+		return nil
+	}
+}
+
+// as g.It's are added conceivably will be with new jenkins job def's
```
Comment (bparees) [inline test/extended/jenkins/jenkins_plugin.go]: scrambled godoc

Comment (bparees) [inline test/extended/jenkins/jenkins_plugin.go]: or scrambled comment.  not sure what it was intended to be.

File: test/extended/jenkins/jenkins_plugin.go (hunk 4 above)
Comment (bparees) [inline test/extended/jenkins/jenkins_plugin.go]: 1) Did this question get resolved (whether it's going to wait forever)?
2) Shouldn't this be checking the frontend-prod deployment, not the frontend deployment?

Comment (bparees) [inline test/extended/jenkins/jenkins_plugin.go]: answered (1) for myself, forgot you reworked waitforadeployment.

**Comment by bparees** [comment]
one nit and a question about the deployment that's being verified.  otherwise lgtm.

File: test/extended/jenkins/jenkins_plugin.go (hunk 9 above)
Comment (gabemontero) [inline test/extended/jenkins/jenkins_plugin.go]: just deleted - no longer pertinent as I recall

File: test/extended/jenkins/jenkins_plugin.go (hunk 4 above)
Comment (gabemontero) [inline test/extended/jenkins/jenkins_plugin.go]: for 2) both frontend and frontend-prod are deployed ... I've added a check for both

**Comment by gabemontero** [comment]
ok ... i've pushed the updates in response to the one nit and question

**Comment by bparees** [comment]
lgtm [merge]

**Comment by gabemontero** [comment]
the  https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8311/ failure came from `hack/../test/cmd/volumes.sh`, prior to the launching of the extended tests

the https://ci.openshift.redhat.com/jenkins/job/merge_pull_requests_origin/4555/ failure came from `hack/../test/cmd/images.sh`

can we re-hit the merge button or force the merge of this new extended test (which passed here: https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/8309/) ?

**Comment by bparees** [comment]
i can keep hitting it, but until we sort out the underlying etcd flake issue, don't hold your breath
[merge]

(I don't want to force merge even though this would be low risk, it's just a bad pattern to get into. we need to fix the merge queue)

**Comment by gabemontero** [comment]
@bparees roger that

**Comment by bparees** [comment]
[merge]

**Comment by deads2k** [comment]
[test]

**Comment by deads2k** [comment]
Gofmt, not a flake

**Comment by gabemontero** [comment]
yep, found 1 more thing to fix last night, failed to gofmt ... just pushed

somebody please hit merge button when get the chance

**Comment by bparees** [comment]
[merge]
